blog/_src/posts/2020-10-15-shallow-mailing-list.md

```diff
@@ -0,0 +1,414 @@
+ Deep does not have the right inspector to learn about the struct fields.
+
+
+### How's transient?
+
+**Shallow** does nothing to enforce the Any type.
+This program runs, and in general Shallow never complains about opaque values.
+
+- - -
+
+
+## Type Inference Installs a Precise Type
+
+Original message : <https://groups.google.com/g/racket-users/c/2X5olKMV3C4/m/mJhsp9ZWBgAJ>
+
+#### On 2020-02-14, John Clements wrote:
+
+> I think I may understand what’s going on here, but a student and I worked on this for quite a while today before I found the problem. 
+>
+> Here’s a program: 
+
+```
+#lang typed/racket 
+
+(define-type Store (Mutable-HashTable Integer Value)) 
+(define-type Value (U Real Boolean String)) 
+
+(define top-store
+  (cast
+    (make-hash (list (cons -1 14) (cons 1 #t) (cons 2 #f)))
+    Store))
+
+(hash-set! top-store 5 1234)
+```
+
+> It fails with this error:
+
+```
+contract violation
+expected: (or/c (and/c byte? positive?) #t #f)
+given: 1234
+in: the values of
+the 3rd conjunct of
+(and/c hash?
+       hash-mutable?
+       (hash/c exact-integer?
+               (or/c (and/c byte? positive?) #t #f)
+               #:immutable #f))
+```
+
+### What's going on?
+
+First off, **Deep** runs fine after swapping `cast` for `ann`.
+
+Second, Typed Racket does try to generalize inferred types for mutable data.
+If the only value in the hash is the byte 14 then Deep also runs.
+
+The problem is that Typed Racket does not generalize the inferred value type
+ (U Byte Boolean) and that cast is a run-time tool for enforcing types.
+Casts create contracts to protect mutable data.
+In this program, there are two contracts:
+ one based on the Store type to protect code that uses the hash,
+ and one based on the inferred type to protect the hash against bad writes.
+That second contract raises the error message.
+
+
+### How's transient?
+
+**Shallow** runs successfully.
+The cast looks for a hash, does not make a contract, and ignores the inferred
+ type going forward.
+
+- - -
+
+
+## Same-Arity Functions in a Case Lambda
+
+Original message : <https://groups.google.com/g/racket-users/c/BDrrgW0axGQ/m/P31NxeGHAAAJ>
+
+
+#### On 2019-07-05, Ryan Kramer wrote:
+
+> In the code below, can `maybe-car` have the given type [....]?
+
+```
+#lang typed/racket
+
+(module untyped racket
+  (provide maybe-car)
+  (define (maybe-car x)
+    (cond
+      [(pair? x) (car x)]
+      [else x])))
+
+(require/typed
+ 'untyped
+ [maybe-car (All (a b) (case->
+                        (-> (Pairof a b) a)
+                        (-> a a)))])
+```
+
+> [Current error:]
+
+```
+Type Checker:
+ Type (All (a b) (case-> (-> (Pairof a b) a) (-> a a)))
+  could not be converted to a contract:
+   function type has two cases of arity 1
+```
+
+### What's going on?
+
+**Deep** tries to enforce the type with a Racket `or/c` contract, but or/c
+ does not allow two function contracts with the same arity because has to
```

Comment:
This sentence needs more love, I think. It’s probably several sentences. See if you can put it more clear.